internal/storage/v2/clickhouse/tracestore/factory.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2025 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package tracestore
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+
+	"github.com/jaegertracing/jaeger/internal/storage/v2/api/tracestore"
+	"github.com/jaegertracing/jaeger/internal/telemetry"
+)
+
+var (
+	_ io.Closer          = (*Factory)(nil)
+	_ tracestore.Factory = (*Factory)(nil)
+)
+
+type Factory struct {
+	config Config
+	telset telemetry.Settings
+	conn   driver.Conn
+}
+
+func NewFactory(ctx context.Context, cfg Config, telset telemetry.Settings) (*Factory, error) {
+	f := &Factory{
+		config: cfg,
+		telset: telset,
+	}
+	opts := &clickhouse.Options{
+		Protocol: getProtocol(f.config.Protocol),
+		Addr:     f.config.Addresses,
+		Auth: clickhouse.Auth{
+			Database: f.config.Database,
+		},
+		DialTimeout: f.config.DialTimeout,
+	}
+	basicAuth := f.config.Auth.Basic.Get()
+	if basicAuth != nil {
+		opts.Auth.Username = basicAuth.Username
+		opts.Auth.Password = string(basicAuth.Password)
+	}
+	conn, err := clickhouse.Open(opts)
+	if err != nil {
+		defer conn.Close()
```

Comment:
if error is returned `conn` could be nil. Normally you just return, you don't try to close it.